source/simulator/helper_functions.cc

```diff
@@ -1694,72 +1686,82 @@ namespace aspect
-          for (unsigned int i=0; i<number_of_reaction_steps; ++i)
-            {
-              // Loop over composition element
-              material_model->fill_additional_material_model_inputs(in_C, solution, fe_values_C, introspection);
-
-              material_model->evaluate(in_C, out_C);
-              heating_model_manager.evaluate(in_C, out_C, heating_model_outputs_C);
-
-              for (unsigned int j=0; j<dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell; ++j)
+          std::vector<std::vector<double>> initial_values_C (quadrature_C.size(),std::vector<double> (introspection.n_compositional_fields));
+          initial_values_C = in_C.composition;
+          std::vector<double> initial_values_T (quadrature_C.size());
+          initial_values_T = in_C.temperature;
+
+          // We have to store all values of the temperature and composition fields in one long vector
+          const unsigned int n_q_points = dof_handler.get_fe().base_element(introspection.base_elements.compositional_fields).dofs_per_cell;
+          const unsigned int n_fields = 1 + introspection.n_compositional_fields;
+          VectorType fields (n_q_points * n_fields);
+
+          for (unsigned int j=0; j<n_q_points; ++j)
+            for (unsigned int f=0; f<n_fields; ++f)
+              if (f==0)
+                fields[j*n_fields+f] = in_C.temperature[j];
+              else
+                fields[j*n_fields+f] = in_C.composition[j][f-1];
+
+          ode.explicit_function = [&] (const double /*time*/,
+                                       const VectorType &y,
+                                       VectorType &ydot)
+          {
+            // Loop over composition element
+            for (unsigned int j=0; j<n_q_points; ++j)
+              for (unsigned int f=0; f<n_fields; ++f)
                 {
-                  for (unsigned int c=0; c<introspection.n_compositional_fields; ++c)
-                    {
-                      // simple forward euler
-                      in_C.composition[j][c] = in_C.composition[j][c]
-                                               + reaction_time_step_size * reaction_rate_outputs_C->reaction_rates[j][c];
-                      accumulated_reactions_C[j][c] += reaction_time_step_size * reaction_rate_outputs_C->reaction_rates[j][c];
-                    }
-                  in_C.temperature[j] = in_C.temperature[j]
-                                        + reaction_time_step_size * heating_model_outputs_C.rates_of_temperature_change[j];
-
-                  if (temperature_and_composition_use_same_fe)
-                    accumulated_reactions_T[j] += reaction_time_step_size * heating_model_outputs_C.rates_of_temperature_change[j];
+                  if (f==0)
+                    in_C.temperature[j]      = y[j*n_fields+f];
+                  else
+                    in_C.composition[j][f-1] = y[j*n_fields+f];
```

Comment:
This, too, could then be places into a local lambda function where then you just have to call `copy_one_vector_into_fields(fields, in_C)` or some such.

Names up for discussion of course.